wandb/sdk/launch/agent/agent.py

```diff
@@ -36,6 +37,77 @@
 _logger = logging.getLogger(__name__)
 
 
+def init_pool_processes(jobs: Dict[str, int], lock: Lock) -> None:
+    global _jobs
+    global _lock
+    _jobs = jobs  # type: ignore
+    _lock = lock  # type: ignore
+
+
+def thread_run_job(
+    launch_spec: Dict[str, Any],
+    job: Dict[str, Any],
+    default_config: Dict[str, Any],
+    api: Api,
+) -> None:
+    project = create_project_from_spec(launch_spec, api)
+    _logger.info("Fetching and validating project...")
+    project = fetch_and_validate_project(project, api)
+    _logger.info("Fetching resource...")
+    resource = launch_spec.get("resource") or "local-container"
+    backend_config: Dict[str, Any] = {
+        PROJECT_SYNCHRONOUS: False,  # agent always runs async
+    }
+
+    backend_config["runQueueItemId"] = job["runQueueItemId"]
+    _logger.info("Loading backend")
+    override_build_config = launch_spec.get("build")
+    override_registry_config = launch_spec.get("registry")
+
+    build_config, registry_config = resolve_build_and_registry_config(
+        default_config, override_build_config, override_registry_config
+    )
+    builder = load_builder(build_config)
+
+    default_runner = default_config.get("runner", {}).get("type")
+    if default_runner == resource:
+        backend_config["runner"] = default_config.get("runner")
+    backend = load_backend(resource, api, backend_config)
+    backend.verify()
+    _logger.info("Backend loaded...")
+    run = backend.run(project, builder, registry_config)
+
+    if not run:
+        return
+    with _lock:  # type: ignore
+        _jobs[run.id] = 1  # type: ignore
```

Comment:
Jobs aren't picklable, so threads aren't able to store them in the shared `_jobs` dict. Instead, we use this to track the set of jobs that are running (`Manager` doesn't have a `set` data type, so we have to use `dict` instead). The jobs themselves are now only being tracked by the individual thread that made them.

Reply:
Actually, this could be a problem. This means we can only have as many runs going as threads on the agent runner machine. Ideally, we'd be able to have a thread kick off a run, not have to wait for the run to complete, and somehow track the run's completion status after the thread terminates, and delete the run from the shared dict when it's complete.

Reply:
This will conflict with the changes to agent.py in https://github.com/wandb/wandb/pull/5011 but I think it should be easy to resolve. The changes made in agent.run in 5011 should just work if we make them in thread_run_job where that stuff has been refactored to